org.eclipse.ui.console/src/org/eclipse/ui/internal/console/IOConsolePartitioner.java

```diff
@@ -1465,4 +1478,65 @@ private void checkPartitions() {
 			Assert.isTrue(knownInputPartitions.isEmpty());
 		}
 	}
+
+	public void showPreviousCommand() throws ExecutionException {
+		String command = _commandHistory.prevCommand();
+		replaceIncompleteUserInputWith(command);
+	}
+
+	public void showNextCommand() throws ExecutionException {
+		String command = _commandHistory.nextCommand();
+		replaceIncompleteUserInputWith(command);
+	}
+
+	/**
+	 * Replace the last input the user entered with the given command. If the last
+	 * user input is not the last partition, we append an input partition.
+	 *
+	 * @param command Replaces the last user input.
+	 * @throws ExecutionException An error occurred attempting to insert history.
+	 */
+	@SuppressWarnings("resource")
+	public void replaceIncompleteUserInputWith(String command) throws ExecutionException {
+		if (!enableCommandHistory) {
+			return;
+		}
+
+		int offset;
+		int length;
+		synchronized (partitions) {
+			IOConsolePartition lastPartition = getLastPartition();
+			// If lastPartition is null, there is no user input.
+			// If lastPartition is read-only, the user has already pressed <enter>.
+			// In either case, we need to create a new one to hold the input text.
+			if (lastPartition == null || lastPartition.isReadOnly()) {
+				lastPartition = new IOConsolePartition(getDocument().getLength(), console.getInputStream());
+				partitions.add(lastPartition);
+				inputPartitions.add(lastPartition);
+			}
+			offset = getLastPartition().getOffset();
```

Comment:
what if getLastPartition() returns null?

Reply:
This is already checked for at the line marked below.
```
			IOConsolePartition lastPartition = getLastPartition();
			// If lastPartition is null, there is no user input.
			// If lastPartition is read-only, the user has already pressed <enter>.
			// In either case, we need to create a new one to hold the input text.
			if (lastPartition == null || lastPartition.isReadOnly()) {     //<=============
				lastPartition = new IOConsolePartition(getDocument().getLength(), console.getInputStream());
				partitions.add(lastPartition);
				inputPartitions.add(lastPartition);
			}
			offset = getLastPartition().getOffset();
			length = getLastPartition().getLength();

```